prover/src/matrix/segments.rs

```diff
@@ -54,31 +56,63 @@ impl<B: StarkField, const N: usize> Segment<B, N> {
     {
         let poly_size = polys.num_rows();
         let domain_size = offsets.len();
+        assert!(domain_size.is_power_of_two());
+        assert!(domain_size > poly_size);
+        assert_eq!(poly_size, twiddles.len() * 2);
+        assert!(poly_offset < polys.num_base_cols());
+
+        // allocate memory for the segment
+        let mut data = unsafe { uninit_vector::<[B; N]>(domain_size) };
+        Self::new_with_buffer(data, polys, poly_offset, offsets, twiddles)
+    }
+
+    /// Instantiates a new [Segment] using the provided data buffer by evaluating polynomials in
+    /// the [ColMatrix] starting at the specified offset.
+    ///
+    /// The offset is assumed to be an offset into the view of the matrix where extension field
+    /// elements are decomposed into base field elements. This offset must be compatible with the
+    /// values supplied into [Matrix::get_base_element()] method.
+    ///
+    /// Evaluation is performed over the domain specified by the provided twiddles and offsets.
+    ///
+    /// # Panics
+    /// Panics if:
+    /// - `poly_offset` greater than or equal to the number of base field columns in `polys`.
+    /// - Number of offsets is not a power of two.
+    /// - Number of offsets is smaller than or equal to the polynomial size.
+    /// - The number of twiddles is not half the size of the polynomial size.
+    /// - Number of offsets is smaller than the length of the data buffer
+    pub fn new_with_buffer<E>(
+        data_buffer: Vec<[B; N]>,
+        polys: &ColMatrix<E>,
+        poly_offset: usize,
+        offsets: &[B],
+        twiddles: &[B],
+    ) -> Self
+    where
+        E: FieldElement<BaseField = B>,
+    {
+        let poly_size = polys.num_rows();
+        let domain_size = offsets.len();
+        let mut data = data_buffer;
 
         assert!(domain_size.is_power_of_two());
         assert!(domain_size > poly_size);
         assert_eq!(poly_size, twiddles.len() * 2);
         assert!(poly_offset < polys.num_base_cols());
+        assert_eq!(data.len(), domain_size);
 
         // determine the number of polynomials to add to this segment; this number can be either N,
         // or smaller than N when there are fewer than N polynomials remaining to be processed
         let num_polys_remaining = polys.num_base_cols() - poly_offset;
         let num_polys = if num_polys_remaining < N {
+            // the segment will remain unfilled so we pad it with zeros
+            data.fill([B::ZERO; N]);
```

Comment:
I would probably move this initialization into `new()` constructor. My thinking is that if we are using `new_with_buffer()`, the caller is responsible for initializing the buffer as they need.